src/v/storage/log_reader.cc

```diff
@@ -382,133 +382,133 @@ void log_reader::maybe_log_load_slice_depth_warning(
+                for (auto& b : batches) {
+                    if (b.base_offset() > _expected_next) {
+                        auto gb = make_ghost_batches(
+                          _expected_next.value(),
+                          model::prev_offset(b.base_offset()),
+                          b.term());
+                        std::move(
+                          gb.begin(),
+                          gb.end(),
+                          std::back_inserter(batches_filled));
+                    }
+                    _expected_next = model::next_offset(b.last_offset());
+                    batches_filled.emplace_back(std::move(b));
+                }
+                co_return std::move(batches_filled);
+            }
+            // To keep things consistent, our internal accounting is all done in
+            // untranslated offsets, even if we've been requested to return
+            // translated offsets.
+            _expected_next = model::next_offset(batches.back().last_offset());
+
+            if (_config.translate_offsets) {
+                vassert(
+                  _translator, "Expected offset translactor to be initialized");
+                for (auto& b : batches) {
+                    b.header().base_offset = _translator->from_log_offset(
+                      b.base_offset());
+                }
+            }
 
-ss::future<log_reader::storage_t>
-log_reader::load_slice(model::timeout_clock::time_point timeout) {
-    _load_slice_depth++;
-    if (is_done()) {
-        // must keep this function because, the segment might not be done
-        // but offsets might have exceeded the read
-        set_end_of_stream();
-        return _iterator.close().then(
-          [] { return ss::make_ready_future<storage_t>(); });
-    }
-    if (_last_base == _config.start_offset) {
-        set_end_of_stream();
-        return _iterator.close().then(
-          [] { return ss::make_ready_future<storage_t>(); });
-    }
-    /**
-     * We do not want to close the reader if we stopped because requested range
-     * was read. This way we make it possible to reset configuration and reuse
-     * underlying file input stream.
-     */
-    if (
-      _config.start_offset > _config.max_offset
-      || _config.bytes_consumed > _config.max_bytes || _config.over_budget) {
-        set_end_of_stream();
-        return ss::make_ready_future<storage_t>();
-    }
-    maybe_log_load_slice_depth_warning("reading more");
-    _last_base = _config.start_offset;
-    ss::future<> fut = find_next_valid_iterator();
-    if (is_end_of_stream()) {
-        return fut.then([] { return ss::make_ready_future<storage_t>(); });
+            co_return std::move(batches);
+        } catch (...) {
+            e = std::current_exception();
+            set_end_of_stream();
+            _probe.batch_parse_error();
+        }
+        if (e) {
+            co_await _iterator.close();
+            co_return log_reader::storage_t{};
```

Comment:
`std::rethrow_exception(e)`?

because

```
.handle_exception([this](std::exception_ptr e) {
          set_end_of_stream();
          _probe.batch_parse_error();
          return _iterator.close().then(
            [e] { return ss::make_exception_future<storage_t>(e); });
      });
```

Reply:
Nice spot